File: src/policy.ts
/**
 * cf-cloak Policy Engine
 *
 * A pure, side-effect-free decision function that determines whether a domain
 * request should be blocked or allowed, given a policy configuration.
 *
 * Priority order (highest → lowest):
 *  1. Temporary allows   — user clicked "allow for now" until expiry timestamp
 *  2. Domain overrides   — explicit per-suffix block/allow rules
 *  3. App overrides      — per-Android-package block/allow rules
 *  4. Category check     — category blocking enabled/disabled
 *  5. Canonical rule     — Light/Extreme mode may observe or block the match
 *  6. Blocklist match    — plain matched domains fall back to defaultAction
 *  7. Not in blocklist   — always allow (never block unlisted domains)
 *
 * Licensed under AGPLv3. See the repository root for full terms.
 */

import type { CanonicalRule, ModeAction } from './ruleset.js'

export type Action = 'block' | 'allow'
export type DecisionEffect = 'block' | 'allow' | 'observe'
export type EnforcementMode = 'light' | 'extreme'

/**
 * A temporary exception that allows a specific domain/suffix until
 * the given epoch-ms timestamp. Used for "allow for now" flows.
 */
export interface TempAllow {
  /** Domain name or suffix to allow (case-insensitive, exact or suffix match). */
  domain: string
  /** Unix epoch milliseconds at which this exception expires. */
  expiresAt: number
}

/**
 * The policy configuration that controls blocking behaviour.
 *
 * All fields except `version` are optional on the wire — the evaluator falls
 * back to the defaults in DEFAULT_POLICY when fields are missing.
 */
export interface Policy {
  /** Schema version — increment when the shape changes. */
  version: number
  /** The active enforcement mode for canonical rules. */
  mode?: EnforcementMode
  /**
   * Action applied when a domain is in the blocklist and no other rule
   * overrides it. Default: 'block'.
   */
  defaultAction: Action
  /**
   * Per-category on/off switches.
   *   true  → category blocking is active (default when key is absent)
   *   false → category is opted-out; matching domains are allowed through
   */
  categoryEnabled: Record<string, boolean>
  /**
   * Per-Android-package-name forced action.
   * Overrides category + default but is overridden by domain overrides and
   * temp-allows.
   */
  appOverrides: Record<string, Action>
  /**
    * Per-domain forced action.
    * A key may match the domain exactly or as a suffix (e.g. "evil.com"
    * matches "www.evil.com"). Exact matches beat suffix matches, and among
    * suffixes the longest match wins. Overridden only by temp-allows.
   */
  domainOverrides: Record<string, Action>
  /**
   * Temporary exceptions — checked first, expire automatically.
   * Array is walked in declaration order; the first non-expired match wins.
   */
  tempAllows: TempAllow[]
}

/** The policy applied when none has been explicitly configured. */
export const DEFAULT_POLICY: Policy = {
  version: 2,
  mode: 'light',
  defaultAction: 'block',
  categoryEnabled: {},
  appOverrides: {},
  domainOverrides: {},
  tempAllows: [],
}

export type DecisionReason =
  | 'temp_allow'
  | 'domain_override'
  | 'system_allowlist'
  | 'app_override'
  | 'category_disabled'
  | 'category_blocked'
  | 'rule_observe'
  | 'rule_block'
  | 'default_block'
  | 'default_allow'

export interface Decision {
  action: Action
  effect: DecisionEffect
  reason: DecisionReason
}

export type MatchedPolicyRule = Pick<
  CanonicalRule,
  | 'domain'
  | 'confidenceScore'
  | 'entityNames'
  | 'matchScope'
  | 'sources'
  | 'categories'
  | 'confidenceTier'
  | 'compatibilityTags'
  | 'reviewNotes'
  | 'lightAction'
  | 'extremeAction'
>

/** Input to the evaluate() function. */
export interface PolicyRequest {
  /** The queried domain, already lowercased. */
  domain: string
  /** Explicit active mode for this evaluation when the caller has already resolved it. */
  mode?: EnforcementMode | null
  /**
   * Preferred canonical match object from the ruleset runtime. When present,
   * evaluation reads category and mode actions from here before consulting the
   * legacy decomposed fields below.
   */
  matchedRule?: MatchedPolicyRule | null
  /** Category from blocklist metadata, or null/undefined if no metadata. */
  category?: string | null
  /** Android package name of the requesting app, or null/undefined if unknown. */
  app?: string | null
  /** True when a shipped system allowlist entry matched this domain. */
  systemAllowlisted?: boolean | null
  /**
   * The blocklist suffix that matched this domain, or null/undefined when the
   * domain is NOT in the blocklist.  When absent, the domain is allowed unless
   * a domain/app override explicitly blocks it.
   */
  matchedSuffix?: string | null
  /** Canonical Light-mode action for the matched rule, when available. */
  lightAction?: ModeAction | null
  /** Canonical Extreme-mode action for the matched rule, when available. */
  extremeAction?: ModeAction | null
}

/**
 * Evaluate a DNS/SNI request against a policy and return the blocking decision.
 *
 * This function is pure — it has no side effects. The VPN service acts on the
 * returned decision (block → NXDOMAIN / RST, allow → forward).
 *
 * @param request  The normalised request to evaluate.
 * @param policy   The active policy (pass DEFAULT_POLICY if none configured).
 * @param now      Current epoch-ms — injectable for deterministic testing.
 */
export function evaluate(
  request: PolicyRequest,
  policy: Policy,
  now = Date.now(),
): Decision {
  const domain = request.domain.toLowerCase()
  const mode = resolveEnforcementMode(request.mode ?? policy.mode)
  const matchedDomain = getMatchedDomain(request)
  const matchedCategory = getMatchedCategory(request)

  // ── 1. Temp allows (highest priority) ────────────────────────────────────
  for (const ta of policy.tempAllows) {
    if (ta.expiresAt > now && domainMatchesSuffix(domain, ta.domain.toLowerCase())) {
      return { action: 'allow', effect: 'allow', reason: 'temp_allow' }
    }
  }

  // ── 2. Domain overrides ───────────────────────────────────────────────────
  const domainOverrideAction = getDomainOverrideAction(domain, policy.domainOverrides)
  if (domainOverrideAction !== null) {
    return { action: domainOverrideAction, effect: domainOverrideAction, reason: 'domain_override' }
  }

  // ── 3. System allowlist ───────────────────────────────────────────────────
  if (request.systemAllowlisted && matchedDomain !== null) {
    return { action: 'allow', effect: 'allow', reason: 'system_allowlist' }
  }

  // ── 4. App overrides ──────────────────────────────────────────────────────
  if (request.app) {
    const override = policy.appOverrides[request.app]
    if (override !== undefined) {
      return { action: override, effect: override, reason: 'app_override' }
    }
  }

  // ── 5. Category + default (applies only when domain IS in blocklist) ──────
  if (matchedDomain !== null) {
    if (matchedCategory) {
      const enabled = policy.categoryEnabled[matchedCategory]
      if (enabled === false) {
        // User opted this category out — allow even though it's in the list
        return { action: 'allow', effect: 'allow', reason: 'category_disabled' }
      }
      if (enabled === true) {
        return { action: 'block', effect: 'block', reason: 'category_blocked' }
      }
      // key absent → fall through to defaultAction
    }

    const matchedRuleAction = getMatchedRuleAction(request, mode)
    if (matchedRuleAction === 'observe') {
      return { action: 'allow', effect: 'observe', reason: 'rule_observe' }
    }
    if (matchedRuleAction === 'block') {
      return { action: 'block', effect: 'block', reason: 'rule_block' }
    }

    return {
      action: policy.defaultAction,
      effect: policy.defaultAction,
      reason: policy.defaultAction === 'block' ? 'default_block' : 'default_allow',
    }
  }

  // ── 5. Domain not in blocklist → allow ───────────────────────────────────
  return { action: 'allow', effect: 'allow', reason: 'default_allow' }
}

function getMatchedRuleAction(
  request: PolicyRequest,
  mode: EnforcementMode,
): ModeAction | null {
  if (mode === 'extreme') {
    return request.matchedRule?.extremeAction ?? request.extremeAction ?? null
  }

  return request.matchedRule?.lightAction ?? request.lightAction ?? null
}

function getMatchedDomain(request: PolicyRequest): string | null {
  return request.matchedRule?.domain ?? request.matchedSuffix ?? null
}

function getMatchedCategory(request: PolicyRequest): string | null {
  return request.matchedRule?.categories[0] ?? request.category ?? null
}

function resolveEnforcementMode(mode: EnforcementMode | null | undefined): EnforcementMode {
  return mode ?? DEFAULT_POLICY.mode ?? 'light'
}

function getDomainOverrideAction(
  domain: string,
  overrides: Record<string, Action>,
): Action | null {
  let bestAction: Action | null = null
  let bestMatchLength = -1
  let bestMatchIsExact = false

  for (const [rawCandidate, action] of Object.entries(overrides)) {
    const candidate = rawCandidate.toLowerCase()
    if (!domainMatchesSuffix(domain, candidate)) {
      continue
    }

    const isExact = domain === candidate
    if (
      bestAction === null ||
      (isExact && !bestMatchIsExact) ||
      (isExact === bestMatchIsExact && candidate.length > bestMatchLength)
    ) {
      bestAction = action
      bestMatchLength = candidate.length
      bestMatchIsExact = isExact
    }
  }

  return bestAction
}

/**
 * Returns true if `domain` equals `suffix` or is a subdomain of it.
 * Both arguments must be pre-lowercased.
 */
function domainMatchesSuffix(domain: string, suffix: string): boolean {
  return domain === suffix || domain.endsWith(`.${suffix}`)
}

File: android/src/main/java/app/choicefirst/eu/cloak/Policy.kt
package app.choicefirst.eu.cloak

import org.json.JSONObject

/**
 * Physical network action returned by the policy evaluator.
 */
enum class PolicyAction { BLOCK, ALLOW }

/**
 * User-visible effect of the decision.
 */
enum class DecisionEffect { BLOCK, ALLOW, OBSERVE }

/**
 * Automatic enforcement mode for canonical rule actions.
 */
enum class EnforcementMode { LIGHT, EXTREME }

/**
 * Mode-specific action attached to a canonical rule.
 */
enum class ModeAction { BLOCK, OBSERVE }

/**
 * A temporary exception that allows a domain/suffix until the given
 * epoch-millisecond timestamp. Mirrors [cf-cloak/src/policy.ts TempAllow].
 */
data class TempAllow(
    val domain: String,
    val expiresAt: Long,
)

data class MatchedPolicyRule(
    val domain: String,
    val confidenceScore: Double? = null,
    val entityNames: List<String> = emptyList(),
    val matchScope: String? = null,
    val registrableDomain: String? = null,
    val categories: List<String> = emptyList(),
    val confidenceTier: String? = null,
    val compatibilityTags: List<String> = emptyList(),
    val reviewNotes: List<String> = emptyList(),
    val lightAction: ModeAction? = null,
    val extremeAction: ModeAction? = null,
)

/**
 * Input to [PolicyEngine.evaluate]. All fields are optional except [domain].
 */
data class PolicyRequest(
    val domain: String,
    val mode: EnforcementMode? = null,
    val matchedRule: MatchedPolicyRule? = null,
    val category: String? = null,
    val app: String? = null,
    val systemAllowlisted: Boolean = false,
    /** The blocklist suffix that matched this domain, or null if no match. */
    val matchedSuffix: String? = null,
    /** Canonical Light-mode action for the matched rule, when available. */
    val lightAction: ModeAction? = null,
    /** Canonical Extreme-mode action for the matched rule, when available. */
    val extremeAction: ModeAction? = null,
)

/**
 * Reason codes returned alongside the [Decision]. Used for logging/analytics.
 */
enum class DecisionReason {
    TEMP_ALLOW,
    DOMAIN_OVERRIDE,
    SYSTEM_ALLOWLIST,
    APP_OVERRIDE,
    CATEGORY_DISABLED,
    CATEGORY_BLOCKED,
    RULE_OBSERVE,
    RULE_BLOCK,
    DEFAULT_BLOCK,
    DEFAULT_ALLOW,
}

/**
 * The decision returned by [PolicyEngine.evaluate].
 */
data class Decision(
    val action: PolicyAction,
    val effect: DecisionEffect,
    val reason: DecisionReason,
)

/**
 * Policy configuration that controls blocking behaviour.
 *
 * Mirrors the TypeScript [Policy] interface in cf-cloak/src/policy.ts.
 * Construct via [Policy.fromJson] when receiving from the Capacitor bridge,
 * or use [Policy.DEFAULT] for the factory-default policy.
 */
data class Policy(
    val version: Int = 2,
    val mode: EnforcementMode = EnforcementMode.LIGHT,
    val defaultAction: PolicyAction = PolicyAction.BLOCK,
    /** category → true means blocking is active; false means opted-out. */
    val categoryEnabled: Map<String, Boolean> = emptyMap(),
    /** Android package name → forced action. */
    val appOverrides: Map<String, PolicyAction> = emptyMap(),
    /** Domain suffix → forced action. */
    val domainOverrides: Map<String, PolicyAction> = emptyMap(),
    val tempAllows: List<TempAllow> = emptyList(),
) {
    companion object {
        /** The default policy (block everything in the blocklist). */
        val DEFAULT = Policy()

        /**
         * Deserialise a [Policy] from a JSON string received via the
         * Capacitor bridge. Unknown keys are silently ignored; malformed values
         * fall back to their defaults.
         */
        fun fromJson(json: String): Policy {
            val obj = JSONObject(json)
            val version = obj.optInt("version", 2)

            val mode = when (obj.optString("mode", "light")) {
                "extreme" -> EnforcementMode.EXTREME
                else       -> EnforcementMode.LIGHT
            }

            val defaultAction = when (obj.optString("defaultAction", "block")) {
                "allow" -> PolicyAction.ALLOW
                else    -> PolicyAction.BLOCK
            }

            val categoryEnabled = mutableMapOf<String, Boolean>()
            obj.optJSONObject("categoryEnabled")?.let { ce ->
                ce.keys().forEach { k -> categoryEnabled[k] = ce.getBoolean(k) }
            }

            val appOverrides = mutableMapOf<String, PolicyAction>()
            obj.optJSONObject("appOverrides")?.let { ao ->
                ao.keys().forEach { k ->
                    appOverrides[k] = when (ao.getString(k)) {
                        "allow" -> PolicyAction.ALLOW
                        else    -> PolicyAction.BLOCK
                    }
                }
            }

            val domainOverrides = mutableMapOf<String, PolicyAction>()
            obj.optJSONObject("domainOverrides")?.let { dom ->
                dom.keys().forEach { k ->
                    domainOverrides[k] = when (dom.getString(k)) {
                        "allow" -> PolicyAction.ALLOW
                        else    -> PolicyAction.BLOCK
                    }
                }
            }

            val tempAllows = mutableListOf<TempAllow>()
            obj.optJSONArray("tempAllows")?.let { arr ->
                for (i in 0 until arr.length()) {
                    val ta = arr.getJSONObject(i)
                    tempAllows.add(
                        TempAllow(
                            domain = ta.getString("domain"),
                            expiresAt = ta.getLong("expiresAt"),
                        ),
                    )
                }
            }

            return Policy(version, mode, defaultAction, categoryEnabled, appOverrides, domainOverrides, tempAllows)
        }
    }
}

/**
 * Pure policy evaluator — the Kotlin mirror of cf-cloak/src/policy.ts `evaluate()`.
 *
 * Priority order (highest → lowest):
 *  1. Temporary allows
 *  2. Domain overrides
 *  3. System allowlist
 *  4. App overrides
 *  5. Category check (only when domain is in the blocklist)
 *  6. Canonical rule action (Light/Extreme observe or block)
 *  7. Default action (only when domain is in the blocklist)
 *  8. Not in blocklist → allow
 */
object PolicyEngine {

    fun evaluate(
        request: PolicyRequest,
        policy: Policy,
        now: Long = System.currentTimeMillis(),
    ): Decision {
        val domain = request.domain.lowercase()
        val mode = request.mode ?: policy.mode
        val matchedDomain = matchedDomain(request)
        val matchedCategory = matchedCategory(request)

        // ── 1. Temp allows ──────────────────────────────────────────────────
        for (ta in policy.tempAllows) {
            if (ta.expiresAt > now && domainMatchesSuffix(domain, ta.domain.lowercase())) {
                return Decision(PolicyAction.ALLOW, DecisionEffect.ALLOW, DecisionReason.TEMP_ALLOW)
            }
        }

        // ── 2. Domain overrides ─────────────────────────────────────────────
        val domainOverride = domainOverrideAction(domain, policy.domainOverrides)
        if (domainOverride != null) {
            return Decision(domainOverride, toEffect(domainOverride), DecisionReason.DOMAIN_OVERRIDE)
        }

        // ── 3. System allowlist ─────────────────────────────────────────────
        if (request.systemAllowlisted && matchedDomain != null) {
            return Decision(PolicyAction.ALLOW, DecisionEffect.ALLOW, DecisionReason.SYSTEM_ALLOWLIST)
        }

        // ── 4. App overrides ────────────────────────────────────────────────
        request.app?.let { app ->
            val override = policy.appOverrides[app]
            if (override != null) {
                return Decision(override, toEffect(override), DecisionReason.APP_OVERRIDE)
            }
        }

        // ── 5. Category + mode-aware rule action + default ──────────────────
        if (matchedDomain != null) {
            matchedCategory?.let { cat ->
                val enabled = policy.categoryEnabled[cat]
                if (enabled == false) return Decision(PolicyAction.ALLOW, DecisionEffect.ALLOW, DecisionReason.CATEGORY_DISABLED)
                if (enabled == true)  return Decision(PolicyAction.BLOCK, DecisionEffect.BLOCK, DecisionReason.CATEGORY_BLOCKED)
                // key absent → fall through to defaultAction
            }

            when (matchedRuleAction(request, mode)) {
                ModeAction.OBSERVE -> return Decision(PolicyAction.ALLOW, DecisionEffect.OBSERVE, DecisionReason.RULE_OBSERVE)
                ModeAction.BLOCK -> return Decision(PolicyAction.BLOCK, DecisionEffect.BLOCK, DecisionReason.RULE_BLOCK)
                null -> Unit
            }

            return Decision(
                policy.defaultAction,
                toEffect(policy.defaultAction),
                if (policy.defaultAction == PolicyAction.BLOCK) DecisionReason.DEFAULT_BLOCK
                else DecisionReason.DEFAULT_ALLOW,
            )
        }

        // ── 6. Domain not in blocklist → allow ──────────────────────────────
        return Decision(PolicyAction.ALLOW, DecisionEffect.ALLOW, DecisionReason.DEFAULT_ALLOW)
    }

    private fun matchedRuleAction(
        request: PolicyRequest,
        mode: EnforcementMode,
    ): ModeAction? = when (mode) {
        EnforcementMode.EXTREME -> request.matchedRule?.extremeAction ?: request.extremeAction
        EnforcementMode.LIGHT -> request.matchedRule?.lightAction ?: request.lightAction
    }

    private fun matchedDomain(request: PolicyRequest): String? =
        request.matchedRule?.domain ?: request.matchedSuffix

    private fun matchedCategory(request: PolicyRequest): String? =
        request.matchedRule?.categories?.firstOrNull() ?: request.category

    private fun domainOverrideAction(
        domain: String,
        overrides: Map<String, PolicyAction>,
    ): PolicyAction? {
        var bestAction: PolicyAction? = null
        var bestMatchLength = -1
        var bestMatchIsExact = false

        for ((rawCandidate, action) in overrides) {
            val candidate = rawCandidate.lowercase()
            if (!domainMatchesSuffix(domain, candidate)) {
                continue
            }

            val isExact = domain == candidate
            if (
                bestAction == null ||
                (isExact && !bestMatchIsExact) ||
                (isExact == bestMatchIsExact && candidate.length > bestMatchLength)
            ) {
                bestAction = action
                bestMatchLength = candidate.length
                bestMatchIsExact = isExact
            }
        }

        return bestAction
    }

    private fun toEffect(action: PolicyAction): DecisionEffect = when (action) {
        PolicyAction.BLOCK -> DecisionEffect.BLOCK
        PolicyAction.ALLOW -> DecisionEffect.ALLOW
    }

    private fun domainMatchesSuffix(domain: String, suffix: String): Boolean =
        domain == suffix || domain.endsWith(".$suffix")
}
